src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -580,7 +582,35 @@ void CoreWorkerDirectTaskSubmitter::PushNormalTask(
           RAY_CHECK_GE(scheduling_key_entry.num_busy_workers, 1u);
           scheduling_key_entry.num_busy_workers--;
 
+          std::thread::id thread_used = std::this_thread::get_id();
           if (!status.ok() || !is_actor_creation || reply.worker_exiting()) {
+            RAY_LOG(DEBUG) << "Getting error from raylet for task " << task_id;
+            
+            auto callback =
+              [this, status, is_actor, task_id, thread_used, addr, reply](const Status &get_task_result_reply_status,
+                     const rpc::GetTaskResultReply &get_task_result_reply) {
+                std::thread::id callback_thread_id = std::this_thread::get_id();
+                RAY_CHECK(thread_used == callback_thread_id) << "Threading issue. Callback must be called on the original calling thread. Original thread id " << thread_used << " callback thread id " << callback_thread_id;
```

Comment:
I was a bit confused about this part. Why do we need to check the thread ID?

Reply:
mostly being defensive, since currently this code runs inline, this verifies it is still using the same thread (and in general it seems that's the current threading model) 

Can remove if we don't want to have these checks, let me know your preference?

Reply:
It would be best if we don't need to make the assumption that the code is using the same thread. I'm still a bit confused why that assumption exists.

Is there a way we can remove both that assumption and the check?